standalone/pruntime/pruntime/src/api_server.rs

```diff
@@ -0,0 +1,246 @@
+    ($rpc: literal, $name: ident, $num: expr) => {
+        #[post($rpc, format = "json", data = "<contract_input>")]
+        fn $name(contract_input: Json<ContractInput>) -> JsonValue {
+            debug!(
+                "{}",
+                ::serde_json::to_string_pretty(&*contract_input).unwrap()
+            );
+
+            let input_string = serde_json::to_string(&*contract_input).unwrap();
+            do_ecall_handle!($num, input_string.as_bytes())
+        }
+    };
+}
+
+macro_rules! proxy_get {
+    ($rpc: literal, $name: ident, $num: expr) => {
+        #[get($rpc)]
+        fn $name() -> JsonValue {
+            let input_string = r#"{ "input": {} }"#.to_string();
+            do_ecall_handle!($num, input_string.as_bytes())
+        }
+    };
+}
+
+macro_rules! proxy {
+    (post, $rpc: literal, $name: ident, $num: expr) => {
+        proxy_post!($rpc, $name, $num)
+    };
+    (get, $rpc: literal, $name: ident, $num: expr) => {
+        proxy_get!($rpc, $name, $num)
+    };
+}
+
+fn read_data(data: Data) -> Option<Vec<u8>> {
+    use std::io::Read;
+    let mut stream = data.open();
+    let mut data = Vec::new();
+    stream.read_to_end(&mut data).ok()?;
+    Some(data)
+}
+
+macro_rules! proxy_bin {
+    ($rpc: literal, $name: ident, $num: expr) => {
+        #[post($rpc, data = "<data>")]
+        fn $name(data: Data) -> JsonValue {
+            let data = match read_data(data) {
+                Some(data) => data,
+                None => {
+                    return json!({
+                        "status": "error",
+                        "payload": "Io error: Read input data failed"
+                    })
+                }
+            };
+            do_ecall_handle!($num, &data)
+        }
+    };
+}
+
+macro_rules! proxy_routes {
+    ($(($m: ident, $rpc: literal, $name: ident, $num: expr),)+) => {{
+        $(proxy!($m, $rpc, $name, $num);)+
+        routes![$($name),+]
+    }};
+}
+
+macro_rules! proxy_bin_routes {
+    ($(($rpc: literal, $name: ident, $num: expr),)+) => {{
+        $(proxy_bin!($rpc, $name, $num);)+
+        routes![$($name),+]
+    }};
+}
+
+#[post("/kick")]
+fn kick() {
+    std::process::exit(0);
+}
+
+#[post("/<method>", data = "<data>")]
+fn prpc_proxy(method: String, data: Data) -> Custom<Vec<u8>> {
+    let path_bytes = method.as_bytes();
+    let data = match read_data(data) {
+        Some(data) => data,
+        None => {
+            return Custom(Status::BadRequest, b"Read body failed".to_vec());
+        }
+    };
+
+    let (status_code, output) = runtime::ecall_prpc_request(path_bytes, &data);
+    if let Some(status) = Status::from_code(status_code) {
+        Custom(status, output)
+    } else {
+        error!("prpc: Invalid status code: {}!", status_code);
+        Custom(Status::ServiceUnavailable, vec![])
+    }
+}
+
+#[get("/dump")]
+fn dump_state() -> anyhow::Result<Stream<PipeReader>> {
+    let (r, writer) = os_pipe::pipe()?;
+    std::thread::spawn(move || {
```

Comment:
I see we create a new thread here. What's the reason behind?

Reply:
The web framework requires a `impl Read` to get a streamed response. But the state serializing (serde/cbor) just want to write into a `impl Writer`. So we spawn a thread to connect them with a `pipe`.